_posts/2018-03-15-complete-introduction-to-nodejs.md

```diff
@@ -0,0 +1,406 @@
+
+Here are som useful npm commands to get you started:
+
+`npm init ` <br>Start a new project. This creates a package.json file that keeps track of the installed modules (if you save them).
+
+`npm install <module_name>` <br>Downloads a module that is registered (by name) on the npm registry.
+To see which modules modules are available, simply visit npm.org and search for whatever you need.
+The downloaded code will be saved in the node_modules directory. (Unless global install)
+
+`npm install <module_name> —save` <br> installs and also updates your project's dependencies in package.json
+
+`npm install <module_name> —save-dev` <br> installs and update's your project's development dependencies in package.json (dependencies that you don't need at runtime, I.e. testing frameworks like jasmin or karma, build frameworks like gulp or webpack, …)
+
+`npm install <module_name> --global` <br> installs the package globally, packages with command line interfaces like gulp-cli, angular-cli are installed globally.
+
+`npm uninstall` <br> uninstalls packages from your project
+
+`npm update` <br> updates your packages
+
+For a full overview of npm commands and further documentation of npm, check out this page: [https://docs.npmjs.com/](https://docs.npmjs.com/)
+
+## Node REPL
+Once you have installed node you can open a terminal window and run node.
+This will return a node REPL where you can run javascript code.  for example<br>
+`function add(a, b){ return a + b } <enter>`<br>
+`add(4, 7) <enter>`<br>
+`// Returns 11`<br>
+To terminate the REPL hit ctrl + c
+
+## Hello World
+Using the REPL can be useful sometimes, but when we want to make some persistent programs,
+we might want to write our code in a file and run the content of that file.
+
+In this example we'll create a `helloworld.js` file
+Create it in your favourite IDE or run `touch helloworld.js` and open it in your IDE or vim/nano/…
+
+To keep it simple for this first project, we'll simply make a program that logs 'hello' and 'world' in your terminal,
+ we'll log them separately just to show you that there's different ways to log data with node.
+
+`helloworld.js`<br>
+
+```javascript
+// process is a global variable that refers to the current node process you are running,
+// it has a stdout property that has a write method which we can call to output data.
+process.stdout.write('hello\n'); // the \n creates a new line in the terminal
+
+// or a bit simpler, the one we are used to from the browser, console.log
+console.log('world');
+```
+
+## Hello World Async
+
+An example of non blocking code
+
+```javascript
+// after 2 secs, print world
+setTimeout( ()=> { console.log('world'); }, 2000);
+
+console.log('hello\n');
+// prints  'hello' first, then 'world'
+```
+
+When you've watched the video about how the V8 Engine works, you'll know why 'hello' get's logged first and 'world' second:
+What happens is:
+- setTimeout is added to the call stack,
```

Comment:
`setTimeout` (backticks)
IMO you can just omit the comma's at the end of the list elements.